src/app/clusters/scenes/SceneTable.h

```diff
@@ -0,0 +1,363 @@
+        SceneGroupID mGroupId = kGlobalGroupSceneId;
+        SceneId mSceneId      = kUndefinedSceneId;
+
+        SceneStorageId() = default;
+        SceneStorageId(EndpointId endpoint, SceneId id, SceneGroupID groupId = kGlobalGroupSceneId) :
+            mEndpointId(endpoint), mGroupId(groupId), mSceneId(id)
+        {}
+        SceneStorageId(const SceneStorageId & storageId) :
+            mEndpointId(storageId.mEndpointId), mGroupId(storageId.mGroupId), mSceneId(storageId.mSceneId)
+        {}
+        CHIP_ERROR Serialize(TLV::TLVWriter & writer) const
+        {
+            TLV::TLVType container;
+            ReturnErrorOnFailure(
+                writer.StartContainer(TLV::ContextTag(kTagSceneStorageIDContainer), TLV::kTLVType_Structure, container));
+
+            ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagSceneEndpointID), static_cast<uint16_t>(this->mEndpointId)));
+            ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagSceneGroupID), static_cast<uint16_t>(this->mGroupId)));
+            ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagSceneID), static_cast<uint8_t>(this->mSceneId)));
+
+            return writer.EndContainer(container);
+        }
+        CHIP_ERROR Deserialize(TLV::TLVReader & reader)
+        {
+            TLV::TLVType container;
+            ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::ContextTag(kTagSceneStorageIDContainer)));
+            ReturnErrorOnFailure(reader.EnterContainer(container));
+
+            ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagSceneEndpointID)));
+            ReturnErrorOnFailure(reader.Get(this->mEndpointId));
+            ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagSceneGroupID)));
+            ReturnErrorOnFailure(reader.Get(this->mGroupId));
+            ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagSceneID)));
+            ReturnErrorOnFailure(reader.Get(this->mSceneId));
+
+            return reader.ExitContainer(container);
+        }
+
+        void Clear()
+        {
+            mEndpointId = kInvalidEndpointId;
+            mGroupId    = kGlobalGroupSceneId;
+            mSceneId    = kUndefinedSceneId;
+        }
+        bool operator==(const SceneStorageId & other)
+        {
+            return (this->mEndpointId == other.mEndpointId && this->mGroupId == other.mGroupId && this->mSceneId == other.mSceneId);
+        }
+        void operator=(const SceneStorageId & other)
+        {
+            this->mEndpointId = other.mEndpointId;
+            this->mGroupId    = other.mGroupId;
+            this->mSceneId    = other.mSceneId;
+        }
+    };
+
+    /// @brief struct used to store data held in a scene
+    struct SceneData
+    {
+        char mName[kSceneNameMax]                = { 0 };
+        size_t mNameLength                       = 0;
+        SceneTransitionTime mSceneTransitionTime = 0;
+        ExtensionFieldSetsImpl mExtensionFieldSets;

```

Comment:
This has a particular implementation by value. This means it's not possible to select the implementation. Suggest templating the SceneTable class on `template <class EFSType>` or otherwise not relying on having this in common data of the base class and only including in implementation. Otherwise, what looks like it is user-providable interface implementation, is actually forced to use given concrete implementations and we pay the cost of the virtual interface flexibility, without the benefit.

Reply:
templated in latest commit.